File: src/models/2026-04-29_xxxx/generate/generate.py
#!/usr/bin/env python3
# Simple Assembly Factory — synthetic data generator.
# Builds an assembly-line factory from a YAML config and writes CSVs.
# Install dependencies: pip install pyyaml
#
# Run:  python generate.py

import os
import random
import csv
from dataclasses import dataclass, field
from typing import Optional
import yaml


@dataclass
class Component:
    id: str
    name: str
    level: int
    is_product: bool


@dataclass
class BomEdge:
    input: str
    output: str
    quantity: int


@dataclass
class Workstation:
    id: str
    name: str
    type: str


@dataclass
class Configuration:
    id: str
    workstation: str
    component: str
    processing_time: float
    setup_time: float
    setup_cost: float
    operating_cost: float


@dataclass
class LayoutEdge:
    origin: str
    destination: str
    capacity: float
    cost: float


def generate_simple_assembly(config_path: str, export_csv: bool = True):
    with open(config_path) as f:
        cfg = yaml.safe_load(f)

    seed = cfg["metadata"].get("seed")
    if seed is not None:
        random.seed(seed)

    # --- BOM ------------------------------------------------------------------
    bom_cfg = cfg["bom"]
    n_products = bom_cfg["n_products"]
    depth = bom_cfg["depth"]
    branch_min, branch_max = bom_cfg["branching"]
    qty_min, qty_max = bom_cfg["quantity"]
    sharing_ratio = bom_cfg.get("sharing_ratio", 0.0)

    assert n_products >= 1, "bom.n_products must be >= 1"
    assert depth >= 1, "bom.depth must be >= 1"
    assert branch_min >= 1, "bom.branching[0] must be >= 1"
    assert branch_max >= branch_min, "bom.branching range invalid"

    components: list[Component] = []
    bom_edges: list[BomEdge] = []
    producible: list[str] = []
    shared_pool: dict[int, list[str]] = {}
    counter: dict[int, int] = {}

    def build_subtree(parent_id: str, parent_level: int):
        if parent_level <= 0:
            return
        child_level = parent_level - 1
        for _ in range(random.randint(branch_min, branch_max)):
            pool = shared_pool.get(child_level, [])
            if pool and random.random() < sharing_ratio:
                child = random.choice(pool)
            else:
                counter[child_level] = counter.get(child_level, 0) + 1
                n = counter[child_level]
                prefix = "RAW" if child_level == 0 else "COMP"
                child = f"{prefix}_L{child_level}_{n}"
                components.append(Component(
                    id=child,
                    name=f"{prefix} L{child_level} #{n}",
                    level=child_level,
                    is_product=False,
                ))
                shared_pool.setdefault(child_level, []).append(child)
                if child_level > 0:
                    producible.append(child)
                build_subtree(child, child_level)
            bom_edges.append(BomEdge(
                input=child,
                output=parent_id,
                quantity=random.randint(qty_min, qty_max),
            ))

    for p in range(1, n_products + 1):
        pid = f"PROD_{p}"
        components.append(Component(id=pid, name=f"Product {p}", level=depth, is_product=True))
        producible.append(pid)
        build_subtree(pid, depth)

    # --- Workstations ---------------------------------------------------------
    ws_cfg = cfg["workstations"]
    n_ws = ws_cfg["count"]
    assert n_ws >= 1, "workstations.count must be >= 1"

    workstations: list[Workstation] = [
        Workstation(id="Inv", name="Inventory",          type="source"),
        Workstation(id="QI",  name="Quality Inspection", type="sink"),
    ]
    assembly_ws: list[str] = []
    for i in range(1, n_ws + 1):
        ws_id = f"WS_{i}"
        workstations.append(Workstation(id=ws_id, name=f"Assembly {i}", type="production"))
        assembly_ws.append(ws_id)

    # --- Configurations -------------------------------------------------------
    ccfg = cfg["configurations"]
    prod_min, prod_max = ccfg["producers_per_component"]
    pt_r = ccfg["processing_time"]
    st_r = ccfg["setup_time"]
    sc_r = ccfg["setup_cost"]
    oc_r = ccfg["operating_cost"]

    assert prod_min >= 1, "configurations.producers_per_component[0] must be >= 1"
    prod_min = min(prod_min, n_ws)
    prod_max = max(prod_min, min(prod_max, n_ws))

    def usample(r):
        return random.uniform(r[0], r[1])

    configurations: list[Configuration] = []
    cfg_idx = 0
    for comp in producible:
        n = random.randint(prod_min, prod_max)
        chosen = random.sample(assembly_ws, n)
        for ws in chosen:
            cfg_idx += 1
            configurations.append(Configuration(
                id=f"CFG_{cfg_idx}",
                workstation=ws,
                component=comp,
                processing_time=usample(pt_r),
                setup_time=usample(st_r),
                setup_cost=usample(sc_r),
                operating_cost=usample(oc_r),
            ))

    # --- Layout ---------------------------------------------------------------
    lay = cfg["layout"]
    topology = lay.get("topology", "parallel")
    cap_r = lay["flow_capacity"]
    cost_r = lay["transport_cost"]
    assert topology in ("parallel", "linear"), "layout.topology must be 'parallel' or 'linear'"

    layout_edges: list[LayoutEdge] = []

    def edge(o: str, d: str):
        layout_edges.append(LayoutEdge(
            origin=o,
            destination=d,
            capacity=float(random.randint(cap_r[0], cap_r[1])),
            cost=usample(cost_r),
        ))

    if topology == "parallel":
        for ws in assembly_ws:
            edge("Inv", ws)
            edge(ws, "QI")
    else:
        edge("Inv", assembly_ws[0])
        for i in range(len(assembly_ws) - 1):
            edge(assembly_ws[i], assembly_ws[i + 1])
        edge(assembly_ws[-1], "QI")

    # --- Self-validate --------------------------------------------------------
    produced = {c.component for c in configurations}
    for comp in producible:
        assert comp in produced, f"Producible component {comp} has no configuration"

    # --- Export CSVs ----------------------------------------------------------
    out_dir = None
    if export_csv:
        rel = cfg["output"]["directory"]
        script_dir = os.path.dirname(os.path.abspath(__file__))
        out_dir = rel if os.path.isabs(rel) else os.path.normpath(os.path.join(script_dir, rel))
        os.makedirs(out_dir, exist_ok=True)

        with open(os.path.join(out_dir, "components.csv"), "w", newline="") as f:
            w = csv.writer(f)
            w.writerow(["ID", "Name", "Level", "IsProduct"])
            for c in components:
                w.writerow([c.id, c.name, c.level, c.is_product])

        with open(os.path.join(out_dir, "bom.csv"), "w", newline="") as f:
            w = csv.writer(f)
            w.writerow(["Input", "Output", "Quantity"])
            for e in bom_edges:
                w.writerow([e.input, e.output, e.quantity])

        with open(os.path.join(out_dir, "workstations.csv"), "w", newline="") as f:
            w = csv.writer(f)
            w.writerow(["ID", "Name", "Type"])
            for ws in workstations:
                w.writerow([ws.id, ws.name, ws.type])

        with open(os.path.join(out_dir, "configurations.csv"), "w", newline="") as f:
            w = csv.writer(f)
            w.writerow(["ID", "Workstation", "Component", "ProcessingTime", "SetupTime", "SetupCost", "OperatingCost"])
            for c in configurations:
                w.writerow([c.id, c.workstation, c.component, c.processing_time, c.setup_time, c.setup_cost, c.operating_cost])

        with open(os.path.join(out_dir, "layout.csv"), "w", newline="") as f:
            w = csv.writer(f)
            w.writerow(["Origin", "Destination", "Capacity", "Cost"])
            for e in layout_edges:
                w.writerow([e.origin, e.destination, e.capacity, e.cost])

    return dict(
        components=components,
        bom_edges=bom_edges,
        workstations=workstations,
        configurations=configurations,
        layout_edges=layout_edges,
        producible=producible,
        out_dir=out_dir,
    )


if __name__ == "__main__":
    script_dir = os.path.dirname(os.path.abspath(__file__))
    result = generate_simple_assembly(os.path.join(script_dir, "..", "config.yaml"))
    print(f"Components:     {len(result['components'])}")
    print(f"BOM edges:      {len(result['bom_edges'])}")
    print(f"Workstations:   {len(result['workstations'])}")
    print(f"Configurations: {len(result['configurations'])}")
    print(f"Layout edges:   {len(result['layout_edges'])}")
    print(f"Exported →      {result['out_dir']}")

File: src/models/legacy_models/2026-04-27_rafael_model/generate.jl
#!/usr/bin/env julia
# Simple Assembly Factory — synthetic data generator.
# Builds an assembly-line factory from a YAML config and writes CSVs.
# Install Julia, then use the following command to instal the necessary packages:
# ] add YAML CSV Random DataFrames
#
# Run:    julia generate.jl
# Use:    include(...); result = generate_simple_assembly("config.yaml")

using YAML, Random, CSV, DataFrames

"""
    generate_simple_assembly(config_path; export_csv=true) -> NamedTuple

Returns `(components, bom_edges, workstations, configurations, layout_edges,
producible, out_dir)`. When `export_csv=true`, also writes 5 CSV files
(`components.csv`, `bom.csv`, `workstations.csv`, `configurations.csv`,
`layout.csv`) to `cfg["output"]["directory"]`.
"""
function generate_simple_assembly(config_path::String; export_csv::Bool=true)
    cfg = YAML.load_file(config_path)
    seed = get(cfg["metadata"], "seed", nothing)
    isnothing(seed) || Random.seed!(seed)

    # --- BOM ---------------------------------------------------------------
    bom = cfg["bom"]
    n_products             = bom["n_products"]
    depth                  = bom["depth"]
    branch_min, branch_max = bom["branching"]
    qty_min, qty_max       = bom["quantity"]
    sharing_ratio          = get(bom, "sharing_ratio", 0.0)

    @assert n_products ≥ 1   "bom.n_products must be ≥ 1"
    @assert depth ≥ 1        "bom.depth must be ≥ 1"
    @assert branch_min ≥ 1   "bom.branching[1] must be ≥ 1"
    @assert branch_max ≥ branch_min "bom.branching range invalid"

    components = Vector{NamedTuple{(:id,:name,:level,:is_product),
                                   Tuple{Symbol,String,Int,Bool}}}()
    bom_edges = Vector{NamedTuple{(:input,:output,:quantity),
                                  Tuple{Symbol,Symbol,Int}}}()
    producible  = Symbol[]
    shared_pool = Dict{Int, Vector{Symbol}}()
    counter     = Dict{Int, Int}()

    function build_subtree!(parent_id::Symbol, parent_level::Int)
        parent_level <= 0 && return
        child_level = parent_level - 1
        for _ in 1:rand(branch_min:branch_max)
            pool = get(shared_pool, child_level, Symbol[])
            if !isempty(pool) && rand() < sharing_ratio
                child = rand(pool)                         # reuse subtree
            else
                counter[child_level] = get(counter, child_level, 0) + 1
                n = counter[child_level]
                prefix = child_level == 0 ? "RAW" : "COMP"
                child  = Symbol("$(prefix)_L$(child_level)_$n")
                push!(components, (id=child, name="$prefix L$child_level #$n",
                                   level=child_level, is_product=false))
                push!(get!(shared_pool, child_level, Symbol[]), child)
                child_level > 0 && push!(producible, child)
                build_subtree!(child, child_level)
            end
            push!(bom_edges, (input=child, output=parent_id,
                              quantity=rand(qty_min:qty_max)))
        end
    end

    for p in 1:n_products
        pid = Symbol("PROD_$p")
        push!(components, (id=pid, name="Product $p", level=depth, is_product=true))
        push!(producible, pid)
        build_subtree!(pid, depth)
    end

    # --- Workstations (all assembly) --------------------------------------
    ws_cfg = cfg["workstations"]
    n_ws   = ws_cfg["count"]
    @assert n_ws ≥ 1 "workstations.count must be ≥ 1"

    workstations = Vector{NamedTuple{(:id,:name,:type),
                                     Tuple{Symbol,String,Symbol}}}()
    push!(workstations, (id=:Inv, name="Inventory",          type=:source))
    push!(workstations, (id=:QI,  name="Quality Inspection", type=:sink))

    assembly_ws = Symbol[]
    for i in 1:n_ws
        id = Symbol("WS_$i")
        push!(workstations, (id=id, name="Assembly $i", type=:production))
        push!(assembly_ws, id)
    end

    # --- Configurations ----------------------------------------------------
    ccfg = cfg["configurations"]
    prod_min, prod_max = ccfg["producers_per_component"]
    pt_r = ccfg["processing_time"]
    st_r = ccfg["setup_time"]
    sc_r = ccfg["setup_cost"]
    oc_r = ccfg["operating_cost"]

    @assert prod_min ≥ 1 "configurations.producers_per_component[1] must be ≥ 1"
    prod_min = min(prod_min, n_ws)
    prod_max = clamp(prod_max, prod_min, n_ws)

    usample(r) = rand() * (r[2] - r[1]) + r[1]

    configurations = Vector{NamedTuple{
        (:id,:workstation,:component,:processing_time,:setup_time,
         :setup_cost,:operating_cost),
        Tuple{Symbol,Symbol,Symbol,Float64,Float64,Float64,Float64}}}()

    cfg_idx = 0
    for comp in producible
        n = rand(prod_min:prod_max)
        for ws in shuffle(assembly_ws)[1:n]
            cfg_idx += 1
            push!(configurations, (
                id=Symbol("CFG_$cfg_idx"), workstation=ws, component=comp,
                processing_time=usample(pt_r), setup_time=usample(st_r),
                setup_cost=usample(sc_r),     operating_cost=usample(oc_r)))
        end
    end

    # --- Layout ------------------------------------------------------------
    # parallel: Inv → every assembly WS → QI  (independent stations)
    # linear:   Inv → WS_1 → WS_2 → … → WS_n → QI  (single chain)
    lay      = cfg["layout"]
    topology = Symbol(get(lay, "topology", "parallel"))
    cap_r    = lay["flow_capacity"]
    cost_r   = lay["transport_cost"]
    @assert topology in (:parallel, :linear) "layout.topology must be 'parallel' or 'linear'"

    layout_edges = Vector{NamedTuple{(:origin,:destination,:capacity,:cost),
                                     Tuple{Symbol,Symbol,Float64,Float64}}}()

    edge!(o, d) = push!(layout_edges, (origin=o, destination=d,
        capacity=Float64(rand(cap_r[1]:cap_r[2])), cost=usample(cost_r)))

    if topology == :parallel
        for ws in assembly_ws
            edge!(:Inv, ws)
            edge!(ws, :QI)
        end
    else  # :linear
        edge!(:Inv, assembly_ws[1])
        for i in 1:(length(assembly_ws) - 1)
            edge!(assembly_ws[i], assembly_ws[i+1])
        end
        edge!(assembly_ws[end], :QI)
    end

    # --- Self-validate -----------------------------------------------------
    # Every producible component must appear as the `component` of ≥1 config.
    produced = Set(c.component for c in configurations)
    for comp in producible
        @assert comp in produced "Producible component $comp has no configuration"
    end

    # --- Export CSVs -------------------------------------------------------
    # Output path: absolute → used as-is; relative → resolved next to this
    # script. So the script + config can be dropped into any folder and the
    # output appears right beside them.
    out_dir = nothing
    if export_csv
        rel = cfg["output"]["directory"]
        out_dir = isabspath(rel) ? rel : normpath(joinpath(@__DIR__, rel))
        mkpath(out_dir)
        CSV.write(joinpath(out_dir, "components.csv"), DataFrame(
            ID        = String.(getfield.(components, :id)),
            Name      =          getfield.(components, :name),
            Level     =          getfield.(components, :level),
            IsProduct =          getfield.(components, :is_product)))
        CSV.write(joinpath(out_dir, "bom.csv"), DataFrame(
            Input    = String.(getfield.(bom_edges, :input)),
            Output   = String.(getfield.(bom_edges, :output)),
            Quantity =          getfield.(bom_edges, :quantity)))
        CSV.write(joinpath(out_dir, "workstations.csv"), DataFrame(
            ID   = String.(getfield.(workstations, :id)),
            Name =          getfield.(workstations, :name),
            Type = String.(getfield.(workstations, :type))))
        CSV.write(joinpath(out_dir, "configurations.csv"), DataFrame(
            ID             = String.(getfield.(configurations, :id)),
            Workstation    = String.(getfield.(configurations, :workstation)),
            Component      = String.(getfield.(configurations, :component)),
            ProcessingTime =          getfield.(configurations, :processing_time),
            SetupTime      =          getfield.(configurations, :setup_time),
            SetupCost      =          getfield.(configurations, :setup_cost),
            OperatingCost  =          getfield.(configurations, :operating_cost)))
        CSV.write(joinpath(out_dir, "layout.csv"), DataFrame(
            Origin      = String.(getfield.(layout_edges, :origin)),
            Destination = String.(getfield.(layout_edges, :destination)),
            Capacity    =          getfield.(layout_edges, :capacity),
            Cost        =          getfield.(layout_edges, :cost)))
    end

    return (components=components, bom_edges=bom_edges,
            workstations=workstations, configurations=configurations,
            layout_edges=layout_edges, producible=producible,
            out_dir=out_dir)
end

# Script entry point.
# `PROGRAM_FILE == @__FILE__` is the canonical check, but on Windows the two
# paths can differ in case and slash direction (`\` vs `/`). Normalize before
# comparing so the entry point fires whether launched from WSL or Windows.
function main_script()
    isempty(PROGRAM_FILE) && return false
    norm(p) = lowercase(normpath(abspath(p)))
    norm(PROGRAM_FILE) == norm(@__FILE__)
    result = generate_simple_assembly(joinpath(@__DIR__, "config.yaml"))
    println("Components:    ", length(result.components))
    println("BOM edges:     ", length(result.bom_edges))
    println("Workstations:  ", length(result.workstations))
    println("Configurations:", length(result.configurations))
    println("Layout edges:  ", length(result.layout_edges))
    println("Exported →     ", result.out_dir)
end

main_script()
    

